examples/sub-class.py

```diff
@@ -29,36 +29,30 @@
         sys.path.insert(0, cmd_subfolder)
     import paho.mqtt.client as mqtt
 
-class MyMQTTClass:
-    def __init__(self, clientid=None):
-        self._mqttc = mqtt.Client(clientid)
-        self._mqttc.on_message = self.mqtt_on_message
-        self._mqttc.on_connect = self.mqtt_on_connect
-        self._mqttc.on_publish = self.mqtt_on_publish
-        self._mqttc.on_subscribe = self.mqtt_on_subscribe
+class MyMQTTClass(mqtt.Client):
 
-    def mqtt_on_connect(self, mqttc, obj, flags, rc):
+    def on_connect(self, mqttc, obj, flags, rc):
```

Comment:
> Is there a way to make it so that the inheritance-based interface doesn't need the repeated argument?

The only way to do this in a way that works in a backwards-compatible way (i.e., so you can tack on a callback via simple attribute assignment) is by using introspection techniques to query the number of arguments that the function takes and making your decision based on that.

But this breaks if you pass a decorated function that is generalized to `*args` and `**kwargs` (as many decorators are).  So this is a very fragile approach.

I agree that `on_connect(self, self, ...)` is ugly, but the solutions to work around that *and* maintain backwards compatibility are worse.  If it's the second `self` that's annoying you, you can always make it a `staticmethod`.  You get rid of the superfluous `self` (since all you use is `mqttc` anyway), but you still have to explicitly pass `self` when called in an instance.  But then anyone making subclasses would have to remember that those callbacks either need an additional `self` or a `@staticmethod` decorator.